### PR TITLE
Keying/Despill effect

### PR DESCRIPTION
**Pull request by unfa**
Useful for improving chroma-keyed footage. Can work for R, G or B channels.
Defaults to green screen.

I made this, becasue I needed it, but it'd be great if others could use that too.

I've tested this and it works - i'ts so simple it's hard to break anything too badly...

### CODE REVIEW

**Comment by unfa** [comment]
@sobotka  Alright, I've done everything I could, but for some reason the shader doesn't work now. Is there a way to compile this and get some errors? Currently I just save the files and restart olive to see if it works - if it doesn't I try to find mistakes blindly. I've found and fixed many already, but still it fails.

**Comment by cgvirus** [comment]
@unfa You should use terminal to open Olive Editor to see the error trace. Not only for Olive but any standard program, whenever you need to code you should use terminal to log your errors. For Linux, open or `cd` your terminal in your installation directory, then:
`./olive-editor`
If you have already loaded the frags, after editing the frag just save it from text editor and reassign the effect deleting the old one. The frag will be updated. No need to restart Olive.

File: effects/internal/despill.frag [hunk 1]
```diff
@@ -7,11 +7,15 @@ uniform sampler2D tex;
...
 uniform vec2 resolution;
 uniform int channel;
+uniform float factor;
 
 const vec2 renderScale = vec2(1,1);
 
 void main(void)
 {
+    
+    factor = factor * 0.01; // convert factor from 0-100 scale to 0-1 scale
```
Comment (alcomposer) [inline effects/internal/despill.frag]: Factor is a read only variable, you need to make a new `float factorOut` and use that from then on:

```
float factorOut;
factorOut = factor * 0.01;
```
Then replace all the other factors with factorOut

**Comment by alcomposer** [comment]
I left a comment in your PR how to fix. Also why is this in `effects/internal`? Should it not be in `shaders`?
If you decided to chance the directory- also make sure to change the location of the COMMON.vert in the XML.

File: effects/internal/despill.frag (hunk 1 above)
Comment (unfa) [reply effects/internal/despill.frag]: I see. Thanks! Fixed!

**Comment by unfa** [comment]
> Why is this effect in `effects/internal`? Should it not be in `shaders`?
> If you decide to chance the directory- also make sure to change the location of the COMMON.vert in the XML.

You're right. I first used the Community-Effects asa template hence my mistake. Fixed.

It works now!

I also want to add a "Balance" factor to allow shifting the weighting of channels used to reproduce the missing one.

**Comment by unfa** [comment]
I've pushed another version whht added Balance parameter and refactored code to use functions instead of repeating the same stuff 3 times.
Easier to debug, and more efficient.

I tested it - works as expected.

**Comment by itsmattkc** [comment]
Merged in dff8db1af25a4ab3f0d580cf1c59fb2ffc3666e3

**Pull request closed without merging**